docs/authentication.md

```diff
@@ -0,0 +1,174 @@
+
+```jsx
+// pages/profile.js
+
+import useUser from '../lib/useUser'
+import Layout from '../components/Layout'
+
+export default Profile = () => {
+  // Fetch the user client-side
+  const { user } = useUser({ redirectTo: '/login' })
+
+  // Server-render loading state
+  if (!user || user.isLoggedIn === false) {
+    return <Layout>Loading...</Layout>
+  }
+
+  // Once the user request finishes, show the user
+  return (
+    <Layout>
+      <h1>Your Profile</h1>
+      <pre>{JSON.stringify(user, undefined, 2)}</pre>
+    </Layout>
+  )
+}
+
+export const getServerSideProps = withSession(async function ({ req, res }) {
+  // Get the user's session based on the request
+  const user = req.session.get('user')
+
+  return {
+    props: { user },
+  }
+})
+```
+
+An advantage of this pattern is preventing a flash of unauthenticated content before redirecting. It's important to note fetching user data in `getServerSideProps` will block rendering until the request to your authentication provider resolves. To prevent creating a bottleneck and decreasing your TTFB ([Time to First Byte](https://web.dev/time-to-first-byte/)), you should ensure your authentication lookup is fast. Otherwise, consider static generation.
+
+There is an [open RFC](https://github.com/vercel/next.js/discussions/14890) to improve redirecting inside `getServerSideProps`.
+
+## Providers
+
+Now that we've discussed authentication patterns, let's look at specific providers and explore how they're used with Next.js.
+
+### Firebase
+
+<details open>
+  <summary><b>Examples</b></summary>
+  <ul>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/with-firebase-authentication">with-firebase-authentication</a></li>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/with-firebase-authentication-ssr">with-firebase-authentication-ssr</a></li>
+  </ul>
+</details>
+
+When using Firebase Authentication], we recommend using the static generation pattern.
+
+It is possible to use the Firebase Client SDK to generate an ID token and forward it directly to Firebase's REST API on the server to log-in. However, requests to Firebase might take some time to resolve, depending on your user's location.
+
+You can either use [FirebaseUI](https://github.com/firebase/firebaseui-web-react) for a drop-in UI, or create your own with a [custom React hook](https://usehooks.com/useAuth/).
+
+
+### Bring Your Own Database
+
+<details open>
+  <summary><b>Examples</b></summary>
+  <ul>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/with-next-auth">with-next-auth</a></li>
+    <li><a href="https://github.com/vercel/next.js/tree/canary/examples/with-iron-session">with-iron-session</a></li>
+  </ul>
+</details>
+
+If you have an existing database with user data, you'll likely want to utilize an open-source solution that's provider agnostic.
+
+- If you need email/password log-in, use `next-iron-session`.
```

Comment:
Thanks for mentioning next-iron-session in many places here, appreciated! One thing I'd like to add is that I use next-iron-session with a slack-only login. Not even user/password. Not sure how we could update this part so next-iron-session is not seen as a email/password only.

any ideas?

Reply:
I'm not sure it makes sense to mention Slack login – it seems like a small use case. Can we explain the pros/cons of `next-iron-session` better for why you might use it over `next-auth`?

Reply:
Following up @vvo – do you have any thoughts?

Reply:
Here's a proposal:

```md
Here's some information to help you decide on using use `next-auth` or `next-iron-session`:

`next-auth` combines sessions, authentication providers (GitHub, Apple and 20+) and user accounts (if you need it) into a single package. Session data can be stored on your server (database) or on the clients (JWT/JWE).

Example usage: You want to add a GitHub login, have it create user accounts and use JWTs as sessions, use `next-auth`.

`next-iron-session` provides `req.session.set/get/unset/save/destroy` methods and that's it. The session data is encrypted and stored on the client as secured, http-only and same-site:lax cookies.

Example usage: You want sessions helpers you can plug wherever you want, and a small API.
```

Do you want me to add this as a suggestion to your document via GitHub?

Side note: next-auth definitely looks like a more complete solution than next-iron-session. But they have different goals and public.

For example, **I** get "scared" the more options I see in a library. Because it usually means it will get into my way at some point.

Others will get scared by the scarcity of packages like next-iron-session, wondering how they could use that to build their GitHub OAuth login for their app.

Let me know!

Reply:
FWIW I love the idea of steering by example use cases so people can find the right solution for their scenario.

Reply:
This PR was just a starting point, happy to further iterate on this auth section. It's long overdue!